### PR TITLE
isPreRelease(), hasBuild(), clean-ups, smaller improvements (SemVer Test Suite successful)

### PR DESCRIPTION
**Pull request by raubv0gel**

### CODE REVIEW

File: src/Version.php [hunk 1]
```diff
@@ -64,7 +64,7 @@ public function __get(string $property)
      */
     public function __toString(): string
     {
-        $version = implode('.', [$this->major, $this->minor, $this->patch]);
+        $version = "{$this->major}.{$this->minor}.{$this->patch}";
```
Comment (raubv0gel) [inline src/Version.php:67]: better performance

File: src/Traits/Incrementable.php [hunk 2]
```diff
@@ -62,7 +65,7 @@ public function incrementPreRelease(): self
             return $this;
         }
 
-        array_push($identifiers, (string) ((int) array_pop($identifiers) + 1));
+        $identifiers[] = (string) ((int) array_pop($identifiers) + 1);
```
Comment (raubv0gel) [inline src/Traits/Incrementable.php:68]: better performance

File: src/Exceptions/InvalidVersionException.php [hunk 3]
```diff
@@ -4,6 +4,9 @@
 
 use Exception;
 
+/**
+ * Exception that is thrown when an invalid semantic version string is used.
+ */
```
Comment (PHLAK) [inline src/Exceptions/InvalidVersionException.php:9]: I'd prefer to omit unnecessary doc blocks and comments.

```suggestion
```

File: src/Traits/Comparable.php [hunk 4]
```diff
@@ -4,6 +4,9 @@
 
 use PHLAK\SemVer\Version;
 
+/**
+ * Trait for comparing two version objects.
+ */
```
Comment (PHLAK) [inline src/Traits/Comparable.php:9]: I'd prefer to omit unnecessary doc blocks and comments.

```suggestion
```

File: src/Support/helpers.php [hunk 5]
```diff
@@ -1,15 +1,16 @@
...
 
+use PHLAK\SemVer\Exceptions\InvalidVersionException;
 use PHLAK\SemVer\Version;
 
 if (! function_exists('semver')) {
     /**
      * Create a SemVer version object.
      *
-     * @throws \PHLAK\SemVer\Exceptions\InvalidVersionException
+     * @throws InvalidVersionException if the provided semantic version string is invalid
```
Comment (PHLAK) [inline src/Support/helpers.php:10]: I'd prefer to omit the description here.

```suggestion
     * @throws InvalidVersionException
```

File: src/Traits/Incrementable.php [hunk 6]
```diff
@@ -2,6 +2,9 @@
 
 namespace PHLAK\SemVer\Traits;
 
+/**
+ * Trait for incrementing a version object.
+ */
```
Comment (PHLAK) [inline src/Traits/Incrementable.php:7]:
```suggestion
```

File: src/Version.php (hunk 1 above)
Comment (PHLAK) [reply src/Version.php:67]: Again, performance is negligible, even over thousands of iterations.

```
> timeit -n 1000 implode('.', ['foo', 'bar', 'baz']);
= "foo.bar.baz"

Command took 0.000001 seconds on average (0.000001 median; 0.000759 total) to complete.

> timeit -n 1000 'foo' . '.' . 'bar' . '.' . 'baz'
= "foo.bar.baz"

Command took 0.000000 seconds on average (0.000000 median; 0.000475 total) to complete.
```

```suggestion
        $version = implode('.', [$this->major, $this->minor, $this->patch]);
```

File: src/Version.php [hunk 7]
```diff
@@ -84,7 +84,7 @@ public function __toString(): string
      *
      * @param string $version Version string
      *
-     * @throws \PHLAK\SemVer\Exceptions\InvalidVersionException
+     * @throws InvalidVersionException if the provided semantic version string is invalid
```
Comment (PHLAK) [inline src/Version.php:87]:
```suggestion
     * @throws InvalidVersionException
```

**Review by PHLAK — CHANGES_REQUESTED** [review]
Thanks for the PR! The core changes look good. However, there's many unnecessary changes included in this PR. If we can get this PR focused on only the necessary changes (i.e. `isPreRelease()` and `hasBuild()` methods + tests) we can get this merged.

File: src/Exceptions/InvalidVersionException.php (hunk 3 above)
Comment (raubv0gel) [reply src/Exceptions/InvalidVersionException.php:9]: I think it’s bad practice not to PHPDoc (public) classes, traits, interfaces, functions. You have to decide. 🙃

File: src/Traits/Comparable.php (hunk 4 above)
Comment (raubv0gel) [reply src/Traits/Comparable.php:9]: I think it’s bad practice not to PHPDoc (public) classes, traits, interfaces, functions. You have to decide. 🙃

File: src/Support/helpers.php (hunk 5 above)
Comment (raubv0gel) [reply src/Support/helpers.php:10]: I think complete PHPDoc is good practice. You have to decide. 🙃

File: src/Traits/Incrementable.php (hunk 6 above)
Comment (raubv0gel) [reply src/Traits/Incrementable.php:7]: I think it’s bad practice not to PHPDoc (public) classes, traits, interfaces, functions. You have to decide. 🙃

File: src/Version.php (hunk 7 above)
Comment (raubv0gel) [reply src/Version.php:87]: I think complete PHPDoc is good practice. You have to decide. 🙃

File: src/Version.php (hunk 1 above)
Comment (raubv0gel) [reply src/Version.php:67]: You can decide. 🙃

File: src/Traits/Incrementable.php (hunk 2 above)
Comment (raubv0gel) [reply src/Traits/Incrementable.php:68]: You can decide. 🙃

File: src/Version.php [hunk 8]
```diff
@@ -109,12 +109,32 @@ public static function parse(string $version): self
...
     }
 
+    /**
+     * Check if the version is a pre-release.
+     *
+     * @return bool true if the version is a pre-release, false otherwise
+     */
+    public function isPreRelease(): bool
+    {
+        return $this->preRelease !== null && $this->preRelease !== '';
```
[the remark this answers is not in the document]
Comment (raubv0gel) [reply src/Version.php:119]: That’s fine!

File: src/Version.php [hunk 9]
```diff
@@ -109,12 +109,32 @@ public static function parse(string $version): self
...
+    }
+
+    /**
+     * Check if the version has a build string.
+     *
+     * @return bool true if the version has a build string, false otherwise
+     */
+    public function hasBuild(): bool
+    {
+        return $this->build !== null && $this->build !== '';
```
[the remark this answers is not in the document]
Comment (raubv0gel) [reply src/Version.php:129]: It’s fine!

**Comment by raubv0gel** [comment]
Thanks for reviewing!

File: src/Version.php (hunk 1 above)
Comment (PHLAK) [reply src/Version.php:67]: As I stated in my comment, let's strip this PR back to just the method + tests changes.

> If we can get this PR focused on only the necessary changes (i.e. `isPreRelease()` and `hasBuild()` methods + tests) we can get this merged.

File: src/Support/helpers.php [hunk 10]
```diff
@@ -1,15 +1,16 @@
 <?php
 
+use PHLAK\SemVer\Exceptions\InvalidVersionException;
```
Comment (PHLAK) [inline src/Support/helpers.php:3]: Let's omit the changes to this file.

File: src/Version.php [hunk 11]
```diff
@@ -38,7 +38,7 @@ class Version
      *
      * @param string $version Version string
      *
-     * @throws \PHLAK\SemVer\Exceptions\InvalidVersionException
+     * @throws InvalidVersionException if the provided semantic version string is invalid
```
Comment (PHLAK) [inline src/Version.php:41]: Let's not include this change in this PR.

File: src/Version.php [hunk 12]
```diff
@@ -84,7 +84,7 @@ public function __toString(): string
      *
      * @param string $version Version string
      *
-     * @throws \PHLAK\SemVer\Exceptions\InvalidVersionException
+     * @throws InvalidVersionException
      *
```
Comment (PHLAK) [inline src/Version.php:88]: Let's not include this change in this PR.

File: src/Version.php [hunk 13]
```diff
@@ -93,7 +93,7 @@ public static function parse(string $version): self
         $semverRegex = '/^v?(?<major>\d+)(?:\.(?<minor>\d+)(?:\.(?<patch>\d+))?)?(?:-(?<pre_release>[0-9A-Za-z-.]+))?(?:\+(?<build>[0-9A-Za-z-.]+)?)?$/';
 
         if (! preg_match($semverRegex, $version, $matches)) {
-            throw new InvalidVersionException('Invalid semantic version string provided');
+            throw new InvalidVersionException('Invalid semantic version string provided.');
```
Comment (PHLAK) [inline src/Version.php:96]: Let's not include this change in this PR.

File: src/Version.php [hunk 14]
```diff
@@ -109,12 +109,32 @@ public static function parse(string $version): self
...
+        return ! empty($this->build);
+    }
+
     /**
      * Set (override) the entire version value.
      *
      * @param string $version Version string
      *
-     * @throws \PHLAK\SemVer\Exceptions\InvalidVersionException
+     * @throws InvalidVersionException if the provided semantic version string is invalid
```
Comment (PHLAK) [inline src/Version.php:137]: Let's not include this change in this PR.

File: src/Version.php [hunk 15]
```diff
@@ -123,7 +143,7 @@ public function setVersion(string $version): self
         $semverRegex = '/^v?(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+)(?:-(?<pre_release>[0-9A-Za-z-.]+))?(?:\+(?<build>[0-9A-Za-z-.]+)?)?$/';
 
         if (! preg_match($semverRegex, $version, $matches)) {
-            throw new InvalidVersionException('Invalid semantic version string provided');
+            throw new InvalidVersionException('Invalid semantic version string provided.');
```
Comment (PHLAK) [inline src/Version.php:146]: Let's not include this change in this PR.

**Comment by PHLAK** [comment]
Implemented in a16851cc529ba30a2ba3a9eb785fa0249afd2c89

**Pull request closed without merging**